### PR TITLE
ToggleSwitch dragging

### PR DESCRIPTION
**Pull request by Splitwirez**
## What does the pull request do?

Allows ToggleSwitches to be dragged to either end to explicitly set their value to True or False.

## What is the current behavior?

Attempting to drag does not drag, and merely toggles the value as if the switch was clicked.

## What is the updated/expected behavior with this PR?

Dragging the thumb on a ToggleSwitch should allow it to be dragged to either end.

## Checklist

- [ ] Added unit tests (if possible)?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

## Breaking changes

Pretty sure there aren't any, I was careful to ensure the existing template would still work.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/AvaloniaUI/Avalonia?pullRequest=4241) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/AvaloniaUI/Avalonia?pullRequest=4241) before we can accept your contribution.<br/><br/>:x: Splitwirez [sign now](https://cla.dotnetfoundation.org/AvaloniaUI/Avalonia?pullRequest=4241)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/AvaloniaUI/Avalonia?pullRequest=4241) it.</sub>

File: src/Avalonia.Controls/ToggleSwitch.cs [hunk 1]
```diff
@@ -131,6 +142,92 @@ protected override bool RegisterContentPresenter(IContentPresenter presenter)
 
             return result;
         }
+
+
+        Panel _knobsPanel;
```
Comment (danwalmsley) [inline src/Avalonia.Controls/ToggleSwitch.cs:147]: please move these fields to the top of the file with the others.

File: src/Avalonia.Controls/ToggleSwitch.cs [hunk 2]
```diff
@@ -131,6 +142,92 @@ protected override bool RegisterContentPresenter(IContentPresenter presenter)
...
+            
+            _knobsPanel.PointerPressed += KnobsPanel_PointerPressed;
+            _knobsPanel.PointerReleased += KnobsPanel_PointerReleased;
+            _knobsPanel.PointerMoved += KnobsPanel_PointerMoved;
+
+            if (IsChecked.HasValue)
+                UpdateKnobPos(IsChecked.Value);
+        }
+
+        Point _switchStartPoint = new Point();
```
Comment (danwalmsley) [inline src/Avalonia.Controls/ToggleSwitch.cs:165]: put fields at the top

File: src/Avalonia.Controls/ToggleSwitch.cs [hunk 3]
```diff
@@ -131,6 +142,92 @@ protected override bool RegisterContentPresenter(IContentPresenter presenter)
...
+                
+                PseudoClasses.Set(":dragging", false);
+
+                if (shouldBecomeChecked == IsChecked)
+                    UpdateKnobPos(shouldBecomeChecked);
+                else
+                    IsChecked = shouldBecomeChecked;
+            }
+            else
+                base.Toggle();
```
Comment (danwalmsley) [inline src/Avalonia.Controls/ToggleSwitch.cs:190]: please used { } around all statemtnes )if and else above too.

File: src/Avalonia.Controls/ToggleSwitch.cs [hunk 4]
```diff
@@ -131,6 +142,92 @@ protected override bool RegisterContentPresenter(IContentPresenter presenter)
...
+                var difference = e.GetPosition(_switchKnob) - _switchStartPoint;
+
+                if ((!_isDragging) && (System.Math.Abs(difference.X) > 3))
+                {
+                    _isDragging = true;
+                    PseudoClasses.Set(":dragging", true);
+                }
+
+                if (_isDragging)
+                    Canvas.SetLeft(_knobsPanel, System.Math.Min(_switchKnob.Bounds.Width, System.Math.Max(0, (_initLeft + difference.X))));
```
Comment (danwalmsley) [inline src/Avalonia.Controls/ToggleSwitch.cs:211]: missing braces.

File: src/Avalonia.Controls/ToggleSwitch.cs [hunk 5]
```diff
@@ -131,6 +142,92 @@ protected override bool RegisterContentPresenter(IContentPresenter presenter)
...
+            if ((_switchKnob != null) && (!_switchKnob.IsPointerOver))
+                base.Toggle();
+        }
+
+        protected void UpdateKnobPos(bool value)
+        {
+            if ((_switchKnob != null) && (_knobsPanel != null))
+            {
+                if (value)
+                    Canvas.SetLeft(_knobsPanel, _switchKnob.Bounds.Width);
```
Comment (danwalmsley) [inline src/Avalonia.Controls/ToggleSwitch.cs:226]: braces

**Review by danwalmsley — CHANGES_REQUESTED** [review]
some minor nits

File: src/Avalonia.Controls/ToggleSwitch.cs [hunk 6]
```diff
@@ -14,6 +14,17 @@ static ToggleSwitch()
         {
             OffContentProperty.Changed.AddClassHandler<ToggleSwitch>((x, e) => x.OffContentChanged(e));
             OnContentProperty.Changed.AddClassHandler<ToggleSwitch>((x, e) => x.OnContentChanged(e));
+            IsCheckedProperty.Changed.AddClassHandler<ToggleSwitch>((x, e) =>
+            {
+                if ((e.NewValue != null) && (e.NewValue is bool val))
+                    x.UpdateKnobPos(val);
+            });
+
+            BoundsProperty.Changed.AddClassHandler<ToggleSwitch>((x, e) =>
```
Comment (MarchingCube) [inline src/Avalonia.Controls/ToggleSwitch.cs:23]: @grokys I think such handlers could be just done in layout arrange, right?

Comment (Splitwirez) [reply src/Avalonia.Controls/ToggleSwitch.cs:31]: @MarchingCube I tried that. No luck.

File: src/Avalonia.Controls/ToggleSwitch.cs [hunk 7]
```diff
@@ -10,10 +10,31 @@ namespace Avalonia.Controls
     /// </summary>
     public class ToggleSwitch : ToggleButton
     {
+        Panel _knobsPanel;
```
Comment (jmacato) [inline src/Avalonia.Controls/ToggleSwitch.cs:13]: define the accessor modifiers like so:
`private Panel _knobsPanel;`

File: src/Avalonia.Controls/ToggleSwitch.cs [hunk 8]
```diff
@@ -131,6 +152,103 @@ protected override bool RegisterContentPresenter(IContentPresenter presenter)
...
+            {
+                base.Toggle();
+            }
+
+            _isDragging = false;
+
+            _knobsPanelPressed = false;
+        }
+
+        bool _isDragging = false;
```
Comment (jmacato) [inline src/Avalonia.Controls/ToggleSwitch.cs:210]: add accessibility modifier+move this on top

**Review by jmacato — CHANGES_REQUESTED** [review]
more nits :)

File: src/Avalonia.Controls/ToggleSwitch.cs (hunk 6 above)
Comment (danwalmsley) [reply src/Avalonia.Controls/ToggleSwitch.cs:32]: you should be able to override ArrangeOverride... and at that point Bounds is valid... then run your code there.

**Review by danwalmsley — APPROVED** [review]
Tested and code is actuallygood.

**Pull request merged**